tests/pulses/test_shape.py

```diff
@@ -0,0 +1,297 @@
+
+
+def test_drag():
+    pulse = Pulse(
+        start=0,
+        duration=50,
+        amplitude=1,
+        frequency=200_000_000,
+        relative_phase=0,
+        shape=Drag(5, 0.2),
+        channel=1,
+        qubit=0,
+    )
+    _if = 0
+
+    assert pulse.duration == 50
+    assert isinstance(pulse.shape, Drag)
+    assert pulse.shape.name == "Drag"
+    assert pulse.shape.rel_sigma == 5
+    assert pulse.shape.beta == 0.2
+    assert repr(pulse.shape) == "Drag(5, 0.2)"
+
+    sampling_rate = 1
+    num_samples = int(pulse.duration / 1 * sampling_rate)
+    x = np.arange(0, num_samples, 1)
+    i = pulse.amplitude * np.exp(
+        -(1 / 2)
+        * (
+            ((x - (num_samples - 1) / 2) ** 2)
+            / (((num_samples) / pulse.shape.rel_sigma) ** 2)
+        )
+    )
+    q = (
+        pulse.shape.beta
+        * (-(x - (num_samples - 1) / 2) / ((num_samples / pulse.shape.rel_sigma) ** 2))
+        * i
+        * sampling_rate
+    )
+
+    np.testing.assert_allclose(pulse.shape.envelope_waveform_i(sampling_rate), i)
+    np.testing.assert_allclose(pulse.shape.envelope_waveform_q(sampling_rate), q)
+
+
+def test_eq():
+    """Checks == operator for pulse shapes."""
+
+    shape1 = Rectangular()
+    shape2 = Rectangular()
+    shape3 = Gaussian(5)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+
+    shape1 = Gaussian(4)
+    shape2 = Gaussian(4)
+    shape3 = Gaussian(5)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+
+    shape1 = GaussianSquare(4, 0.01)
+    shape2 = GaussianSquare(4, 0.01)
+    shape3 = GaussianSquare(5, 0.01)
+    shape4 = GaussianSquare(4, 0.05)
+    shape5 = GaussianSquare(5, 0.05)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+    assert not shape1 == shape4
+    assert not shape1 == shape5
+
+    shape1 = Drag(4, 0.01)
+    shape2 = Drag(4, 0.01)
+    shape3 = Drag(5, 0.01)
+    shape4 = Drag(4, 0.05)
+    shape5 = Drag(5, 0.05)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+    assert not shape1 == shape4
+    assert not shape1 == shape5
+
+    shape1 = IIR([-0.5, 2], [1], Rectangular())
+    shape2 = IIR([-0.5, 2], [1], Rectangular())
+    shape3 = IIR([-0.5, 4], [1], Rectangular())
+    shape4 = IIR([-0.4, 2], [1], Rectangular())
+    shape5 = IIR([-0.5, 2], [2], Rectangular())
+    shape6 = IIR([-0.5, 2], [2], Gaussian(5))
+    assert shape1 == shape2
+    assert not shape1 == shape3
+    assert not shape1 == shape4
+    assert not shape1 == shape5
+    assert not shape1 == shape6
+
+    shape1 = SNZ(5)
+    shape2 = SNZ(5)
+    shape3 = SNZ(2)
+    shape4 = SNZ(2, 0.1)
+    shape5 = SNZ(2, 0.1)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+    assert not shape1 == shape4
+    assert not shape1 == shape5
+
+    shape1 = eCap(4)
+    shape2 = eCap(4)
+    shape3 = eCap(5)
+    assert shape1 == shape2
+    assert not shape1 == shape3
+
+
+def test_demodulation():
```

Comment:
no tests for modulation?

Reply:
It is also testing modulation. It's just testing the mutual consistency.

I usually write (de)modulation, but it's not a valid identifier. Alternatives:
- `_de_modulation`
- `_modem`
- `_modulation` (intended in a generic sense)

Reply:
Well, the tests in this function do not seem to test much. For example, the following implementations of `modulate` and `demodulate` pass all the tests, but they do neither modulation nor demodulation.
```
def modulate(envelope, freq, rate=1, phase=0.0):
    return envelope


def demodulate(modulated, freq, rate=1):
    return modulated / np.mean(modulated)
```
Proper unit tests should test each unit separately.

Reply:
Well, unit tests are meaningful for as long as they test invariants and limit cases. Testing values is not meaningful on its own.

The invariants are not necessarily selecting a single and unique function. In the best case, they do. But testing something is better than testing nothing.

Instead, what I'm attempting to avoid, it's testing too much. I.e. the point in which tests
- are sensitive to the implementation, such that every time you change the implementation you should also change the tests
- encode most of the implementation, essentially repeating it (that's a regression test, not a unit test)

Reply:
Here the feature tested is mainly that they are inverse to each other. Identity is also inverse to itself, obviously.
But a bug could possibly "imbalance" them. That's what the test is preventing.

If you can suggest something that tests what `modulate` means, without reimplementing modulation, you're more than welcome :)

Reply:
Well, for notable integer single-valued functions of a single variable, I perfectly agree with you.

However, even in that case, I believe it would be even better to have:
```py
assert square(5) == 25
assert square(7) == 49

assert square_root(square(5)) == 5
assert square_root(square(7)) == 7
```
because the defining property of the `square_root` is exactly being the inverse of the `square`.

In this case, instead, I see no simple notable value I could test (of course I could test the modulation of a 0 envelope, but this is not going to add much...).
That's why I asked you to provide an actual example. Not because I'm not aware of any other possible way of testing, but because exactly what you wrote was not a viable option for me.

In which independent way would you compute the result of a frequency modulation (say with `freq=0.15`) of a Gaussian envelope?
I would do the exact same of what is written in the function. And if I encode the value, then I would also depend on the other parameters (`rate` and `phase`), and I should decide a certain precision for floats comparison (which could be too loose, or too strict, but this often depends on the situation).

Reply:
In the end: I was struggling deciding which is a relevant way to `test_modulate` in isolation.

I tested `modulate` against the former version of itself. And I believe that has been tested on electronics (i.e. checking that software modulation results were compatible with hardware-modulated ones), which is highly suboptimal for such a simple function.

> Sure. When you implemented `modulate`, how did you make sure that you have implemented the correct thing? Turn those checks into unit tests.

So, I'm sorry, but this is not very *practical*. It's a generic suggestion, though very sensible. I would have applied it, if I could :)
In case you want to help, the more explicit request it: could you provide an implementation of `test_modulate` not subject to the issues I found? (i.e. either Python code or something that I could turn in Python code without other background information)

Reply:
I see. You can still convert your checks (against the former version of itself) into a couple of hardcoded unit tests by following these steps:

- Take a couple of somewhat representative examples of modulation, e.g. Square/Gaussian envelope on some freq.
- Get the modulation result from the old function.
- Make it a hardcoded unit test for the new function.
- Repeat the same for some differing values of the rest of the arguments.

Of course this relies on the assumption that the original implementation was correct, but I think we can afford it. Ideally whoever implemented the original version and validated against hardware and/or pen-and-paper calculation should have written these unit tests, but we have what we have.

Reply:
Done in fb0ef1251af34716d9159f4a871b5e26f232beb1

I'm still against this approach, since we're essentially mixing unit tests with regression tests. Regression could be valuable, but I'd rather set up separately, and properly store data (there is also a suitable Pytest plugin for this https://github.com/ESSS/pytest-regressions).

However, there was enough consensus to do it, and this was the quickiest way (I'd rather limit the time spent on a feature that is not so used), so if you don't have any further comment, I'd resolve the conversation.

Reply:
It is a unit test for sure. It tests if for given input the unit (the `modulate` function) produces expected output. I believe the part that looks like regression to you is how we came up with the test. Namely we used the previous implementation to come up with a test for the current implementation. That's basically the consensus we came to in order to avoid pen and paper calculations and comparison with some hardware modulation. If you forget where the test came from, then it is just a simple unit tests.

Reply:
Well, if you wish *regression* is an orthogonal feature. *unit* is opposite to *integration*, but both of them could be looking for regressions.